### DISCUSSION TITLE
What's the best way to get an interactive debugger in a template?

### DISCUSSION THREAD

**Discussion by gone** [pallets/jinja#1727] [Q&A]
something similar to https://django-extensions.readthedocs.io/en/latest/debugger_tags.html

**Comment by davidism** [comment]
Add Python's `breakpoint` function to the env, then call it.

```python
env.globals["breakpoint"] = breakpoint
```

```jinja
{{ breakpoint() }}
```

When using PyCharm (possible Professional only) you can set breakpoints in templates without any extra setup.

**Reply by gone** [reply]
I tried that and found myself in jinja's guts, without access to the current scope. I'll try again and see if I was doing something wrong

**Reply by davidism** [reply]
The template file itself doesn't exist at runtime, it's compiled to Python. You have access to the Jinja runtime in its frames.

This is true for that Django tag as well, all it's doing is starting a debugger with a breakpoint. You don't have access to the template, you have access to the runtime evaluating the template.

PyCharm has special handling to retranslate that, but it's not something that any other tool would know about, nor something that Jinja could tell them about.
